### PR TITLE
MGMT-4529 convert host and cluster metrics to ELK

### PR DESCRIPTION
**Pull request by slaviered**
- Fix performance issue Bugzilla 1926329
    - convert: Duration, Host stage duration, Host metrics

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -31,7 +31,6 @@ require (
 	github.com/golang/mock v1.5.0
```
Comment (omertuc) [inline go.mod:31]: Were these `go.mod` changes committed on purpose?

File: internal/metrics/metricsManager.go [hunk 2]
```diff
@@ -124,10 +128,13 @@ type MetricsManager struct {
 	serviceLogicFilesystemUsagePercentage              *prometheus.GaugeVec
 }
 
-func NewMetricsManager(registry prometheus.Registerer) *MetricsManager {
+var _ API = &MetricsManager{}
```
Comment (omertuc) [inline internal/metrics/metricsManager.go:131]: What is this line for?

File: go.mod (hunk 1 above)
Comment (slaviered) [reply go.mod:31]: I ran go mod tidy. should I remove the go.mod?

File: internal/metrics/metricsManager.go (hunk 2 above)
Comment (slaviered) [reply internal/metrics/metricsManager.go:131]: It a golang best practice to ensure MetricsManager always adhere to metric.API interface.

File: go.mod (hunk 1 above)
Comment (omertuc) [reply go.mod:31]: It added a `google.golang.org/appengine v1.6.6` dependency which doesn't seem necessary, even with your new code

**Comment by ybettan** [comment]
Can you please add an explanatory commit message to explain why it is important?

**Comment by ybettan** [comment]
Don't we have tests for that PR? How do you make sure all the events are generated properly? and more important that no one will break it in future changes?

File: internal/metrics/metricsManager.go [hunk 3]
```diff
@@ -305,6 +311,7 @@ func NewMetricsManager(registry prometheus.Registerer) *MetricsManager {
 }
 
 func (m *MetricsManager) ClusterRegistered(clusterVersion string, clusterID strfmt.UUID, emailDomain string) {
+	//MGMT-4526 TODO: remove this scrap after ELK dashboards are verified
```
Comment (ybettan) [inline internal/metrics/metricsManager.go:314]: Where is the transfer to ELK? Why don't you generate an event instead?

File: internal/metrics/metricsManager.go [hunk 4]
```diff
@@ -377,8 +389,10 @@ func (m *MetricsManager) ReportHostInstallationMetrics(log logrus.FieldLogger, c
 			diskType = boot.DriveType
 		}
 		switch currentStage {
-		case models.HostStageDone, models.HostStageFailed:
-			m.handleHostInstallationComplete(log, clusterVersion, clusterID, emailDomain, roleStr, hwVendor, hwProduct, diskType, installationStageStr, h)
+		case models.HostStageDone:
+		case models.HostStageFailed:
```
Comment (ybettan) [inline internal/metrics/metricsManager.go:393]: You changed the logic here, the old implementation will enter the case if one of the conditions is true, the new implementation have an empty case for `HostStageDone` (won't do anything in this case).

This is what you meant?
https://play.golang.org/p/eMPuDQ2GFqb

File: internal/metrics/metricsManager.go (hunk 3 above)
Comment (slaviered) [reply internal/metrics/metricsManager.go:314]: We decided that we allow few weeks for the dashboard to run on ELK, monitor them and debug them, and only then cancel the reports to Prometheus. I ma marking the places for deletion now with TODO label so I won't need to do the research again

File: internal/metrics/metricsManager.go (hunk 4 above)
Comment (slaviered) [reply internal/metrics/metricsManager.go:393]: Nope. Will Fix

File: internal/metrics/metricsManager.go [hunk 5]
```diff
@@ -389,46 +403,76 @@ func (m *MetricsManager) ReportHostInstallationMetrics(log logrus.FieldLogger, c
...
 	}
 }
 
-func (m *MetricsManager) handleHostInstallationComplete(log logrus.FieldLogger, clusterVersion string, clusterID strfmt.UUID, emailDomain string,
+func (m *MetricsManager) reportHostMetricsOnInstallationComplete(ctx context.Context, clusterVersion string, clusterID strfmt.UUID, emailDomain string,
 	roleStr string, hwVendor string, hwProduct string, diskType string, installationStageStr string, h *models.Host) {
+	log := logutil.FromContext(ctx, logrus.New())
+
+	//increment the count of successful installed hosts
+	//MGMT-4526 TODO: remove this scrap after ELK dashboards are verified
```
Comment (ybettan) [inline internal/metrics/metricsManager.go:421]: Same here. Why are you removing the metrics without adding ELK event instead?

File: internal/metrics/metricsManager.go (hunk 3 above)
Comment (ybettan) [reply internal/metrics/metricsManager.go:314]: Yes this I understand but you "future deleted metrics" but didn't add any events instead.
Also as I mentioned https://github.com/openshift/assisted-service/pull/1560#issuecomment-831256621 I think we should have some tests for it and not just checking manually in elastic.

Comment (slaviered) [reply internal/metrics/metricsManager.go:314]: These are special cases where we already have measurement in ELK on those metrics based on existing events. So, In these cases we just need to remove the special metric reports to Prometheus without adding a thing

Comment (ybettan) [reply internal/metrics/metricsManager.go:314]: Ack.

File: internal/metrics/metricsManager.go [hunk 6]
```diff
@@ -91,17 +94,17 @@ type API interface {
 	ClusterValidationChanged(clusterVersion string, emailDomain string, clusterValidationType models.ClusterValidationID)
 	InstallationStarted(clusterVersion string, clusterID strfmt.UUID, emailDomain string, userManagedNetworking string)
 	ClusterHostInstallationCount(clusterID strfmt.UUID, emailDomain string, hostCount int, clusterVersion string)
-	ClusterHostsNTPFailures(clusterID strfmt.UUID, emailDomain string, hostNTPFailureCount int)
```
Comment (gamli75) [inline internal/metrics/metricsManager.go:94]: why did you remove this method?

Comment (slaviered) [reply internal/metrics/metricsManager.go:94]: It is not used anywhere. I consulted the original author and he conceded too that it should be removed because it is not required any more

**Review by gamli75 — APPROVED** [review]

**Comment by ybettan** [comment]
/hold
Wait for:
* commit message update
* fix switch case

You can unhold once ready @slaviered

**Review by gamli75 — APPROVED** [review]
